### DISCUSSION TITLE
<script setup>语法的透传问题

### DISCUSSION THREAD

**Discussion by gyt95** [NervJS/taro#12538] [Q & A]
```
// 父组件
<child text="xxx" @success="success"></child>

// 子组件
<template>
  <grand-child />
</template>
<script lang="ts">
export default {
  inheritAttrs: false
}
</script>

<script lang="ts" setup>
// ...
</script>

// 孙组件
<template>
  <nut-xxx :title="text" @success="success"/>
</template>
<script lang="ts" setup>
const attrs = useAttrs();
const { text, success } = attrs;
</script>
```

像以上这样写，属性text没问题，但事件success则无效，如何解决？是Taro版本还未支持还是我哪里写错了

**Comment by b2nil** [comment]
从 `attrs` 中获取事件的时候，应该要用 `onSuccess`：
```ts
<script lang="ts" setup>
  const attrs = useAttrs();
  const { text, onSuccess } = attrs;
</script>
```

具体可查看： [Fallthrough Attributes](https://vuejs.org/guide/components/attrs.html#disabling-attribute-inheritance)
> A `v-on` event listener like `@click` will be exposed on the object as a function under `$attrs.onClick`.

**Reply by gyt95** [reply]
试过了，也不行。然后我还特意内部定义个事件，发现是undefined。。。

```
<template>
  <span @success="handleSuccess"></span>
</template>
<script lang="ts" setup>
const attrs = useAttrs(); 
const { text, onSuccess } = attrs; 

const handleSuccess = (item) => {
  console.log(item);
  console.log(onSuccess);  // undefined
}
</script>
```

**Reply by b2nil** [reply]
上一级组件没有写事件

**Reply by gyt95** [reply]
很奇怪，我在孙组件的const { text, onSuccess } = attrs;后立即打印attrs能看到属性和函数，但是在handleSuccess里打印attrs，则看不到。

请问上一级组件如何写事件？透传也要写吗

**Reply by b2nil** [reply]
你这里 `<span @success="handleSuccess"></span>` 把 `@success` 改成 `@click` 就能看到了。

关键是：

如果你希望从父组件直接透传到孙组件，那么子组件可以直接继承父组件透传的属性，即设置：`inheritAttrs: true`， 或不用写 `inheritAttrs`。

如果你需要在子组件中过滤一下传下来的属性，即设置 `inheritAttrs: false`，再传给孙组件，那么需要手动获取父组件传下来的属性，再绑定到子组件上：

```ts
// 父组件
<child text="xxx" @success="success"></child>

// 子组件
<template>
  <grand-child @success="onSuccess"/>
</template>
<script lang="ts">
export default {
  inheritAttrs: false
}
</script>

<script lang="ts" setup>
const attrs = useAttrs()
const { onSuccess } = attrs
</script>

// 孙组件
<template>
  <nut-xxx :title="text" @success="onSuccess"/>
</template>
<script lang="ts" setup>
const attrs = useAttrs();
const { text, onSuccess } = attrs;
</script>
```

**Reply by gyt95** [reply]
“你这里 <span @success="handleSuccess"></span> 把 @success 改成 @click 就能看到了。”

但是我的组件是nutui的upload组件，本来就是@success，可能我回复时没改过来。除了@success还要@delete和@failure

“如果你希望从父组件直接透传到孙组件，那么子组件可以直接继承父组件透传的属性，即设置：inheritAttrs: true， 或不用写 inheritAttrs。”

其实我就是想透传，不用在子组件里再写一遍，但现在去改inheritAttrs: true也无效，就是孙组件触发事件无法触发透传的事件。如果像你的代码段的写法，那跟用emits差不多。

**Reply by b2nil** [reply]
设置 `inheritAttrs: true` 后，用 `useAttrs` 获取到的属性里面如果有 `onSuccess` 属性，那透传就没问题。
触发不了的话，那肯定是别的问题。
